modules/essensBidAdapter.js

```diff
@@ -0,0 +1,165 @@
+const CONSTANTS = require('src/constants.json')
+const utils = require('src/utils.js')
+const bidfactory = require('src/bidfactory.js')
+const bidmanager = require('src/bidmanager.js')
+const adloader = require('src/adloader')
+const Adapter = require('src/adapter.js')
+const adaptermanager = require('src/adaptermanager');
+
+// Essens Prebid Adapter
+function EssensAdapter () {
+  let baseAdapter = Adapter.createNew('essens')
+
+  const ENDPOINT = 'bid.essrtb.com/bid/prebid_call'
+
+  baseAdapter.callBids = function (bidRequest) {
+    const bids = bidRequest.bids || []
+
+    const essensBids = bids
+      .filter(bid => isPlacementBidComplete(bid))
+      .map(bid => {
+        let essensBid = {}
+        essensBid.impressionId = bid.bidId
+        essensBid.sizes = utils.parseSizesInput(bid.sizes)
+        essensBid.placementId = bid.params.placementId
+
+        if (bid.params.dealId) {
+          essensBid.deal = bid.params.dealId
+        }
+
+        if (bid.params.floorPrice) {
+          essensBid.floorPrice = bid.params.floorPrice
+        }
+
+        return essensBid
+      })
+
+    const bidderRequestId = bidRequest.bidderRequestId
+
+    const cur = ['USD']
+
+    const urlParam = utils.getTopWindowUrl()
+    const uaParam = getUa()
+
+    if (!utils.isEmpty(essensBids)) {
+      const payloadJson = {bidderRequestId: bidderRequestId, cur: cur, url: urlParam, ua: uaParam, imp: essensBids}
+
+      const scriptUrl = '//' + ENDPOINT + '?callback=$$PREBID_GLOBAL$$.essensResponseHandler' +
+        '&bid=' + encodeURIComponent(JSON.stringify(payloadJson))
+      adloader.loadScript(scriptUrl)
+    }
+
+    function isPlacementBidComplete (bid) {
+      if (bid.bidId && bid.params && bid.params.placementId) {
+        return true
+      } else {
+        utils.logError('bid requires missing essential params for essens')
+      }
+    }
+  }
+
+
+  function getUa () {
+    return window.navigator.userAgent
+  }
+
+  $$PREBID_GLOBAL$$.essensResponseHandler = function (essensResponse) {
+    utils.logInfo('received bid request from Essens')
+    if (!isValidResponse(essensResponse)) {
+      const allRequests = getEssesnBidRequests()
+      allRequests.bids.forEach(registerEmptyResponse)
+      return;
+    }
+
+    registerBids(essensResponse);
+
+    function getEssesnBidRequests () {
+      return $$PREBID_GLOBAL$$
+        ._bidsRequested.find(bidSet => bidSet.bidderCode === 'essens')
```

Comment:
I think it's better that you keep tabs on the bid request in the adapter (it's passed to callbids) rather than poking around at private Prebid.js variables which could change.